File: apps/web/src/features/aurora/go-look.ts
/**
 * go-look.ts — Pinned "go look" score implementation
 *
 * Matches contracts/tests/formulas.py go_look() exactly. This is a
 * temporary TypeScript re-implementation of the pinned contract semantics;
 * the canonical implementation lives in crates/helio-core (WASM).
 *
 * Do NOT tune constants — changing any constant requires regenerating
 * golden vectors in contracts/fixtures/vectors/golook.json.
 */

/**
 * Darkness factor for aurora visibility.
 * PINNED: clamp((−6 − sun_alt_deg) / 12, 0, 1)
 * => 0 at civil twilight or brighter (alt >= −6), 1 at astronomical dark (alt <= −18),
 * linear ramp between.
 */
export function darknessFactor(sunAltDeg: number): number {
  return clamp((-6 - sunAltDeg) / 12, 0, 1);
}

function clamp(x: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, x));
}

// ----------------------------------------------------------------
// GoLookInputs — matches the WASM crate's GoLookInputs struct
// ----------------------------------------------------------------

export interface GoLookInputs {
  /** Visible-aurora probability 0..1 at the user's location, sampled from the
   * delay-corrected OVATION grid. */
  ovalVisibleProb: number;
  sunAltDeg: number;
  moonAltDeg: number;
  /** Illuminated fraction of the lunar disk, 0..1. */
  moonIllumFrac: number;
  /** Multi-model consensus means, 0..1 (Open-Meteo, spec §3.6). */
  cloudTotalConsensus: number;
  /** Low cloud weighted heaviest — it's what actually blocks aurora. */
  cloudLowConsensus: number;
  /** Cross-model disagreement 0..1 (0 = all models agree). */
  cloudModelSpread: number;
  /** GOES CSM Tier-0 point answer, 0..1 clear; null when the leg is unavailable. */
  satelliteClearNow: number | null;
}

export type Verdict = "Likely" | "Possible" | "Unlikely";
export type Limiter =
  | "Daylight"
  | "Oval"
  | "CloudObserved"
  | "CloudForecast"
  | "Moon";

export interface GoLookScore {
  score: number;
  verdict: Verdict;
  confidence: number;
  /** The factor that most limits tonight — drives "why it might be wrong" (spec §8.6). */
  dominantLimiter: Limiter;
}

const LIMITER_ORDER: Limiter[] = [
  "Daylight",
  "Oval",
  "CloudObserved",
  "CloudForecast",
  "Moon",
];

/**
 * The on-device "go look" score. v1.0 heuristic.
 *
 * PINNED formulas (match formulas.py go_look() exactly):
 *   darkness     = darkness_factor(sun_alt_deg)
 *   moon_factor  = 1 − 0.6 · moon_illum_frac · clamp(sin(moon_alt_deg·π/180), 0, 1)
 *   clear_fcst   = clamp(1 − (0.7·cloud_low + 0.3·cloud_total), 0, 1)
 *   clear        = satellite_clear_now is null ? clear_fcst : 0.5·clear_fcst + 0.5·sat
 *   score        = oval_visible_prob · darkness · moon_factor · clear
 *   confidence   = (1 − cloud_model_spread) · (sat leg missing ? 0.85 : 1.0)
 *   verdict      = score >= 0.30 → Likely; >= 0.10 → Possible; else Unlikely
 *   limiter      = argmin over factors {Daylight: darkness, Oval: oval_visible_prob,
 *                   CloudObserved: sat-or-1.0, CloudForecast: clear_fcst, Moon: moon_factor}
 *                   ties broken by enum order.
 */
export function goLook(inputs: GoLookInputs): GoLookScore {
  const darkness = darknessFactor(inputs.sunAltDeg);
  const moonFactor =
    1.0 -
    0.6 *
      inputs.moonIllumFrac *
      clamp(Math.sin(toRadians(inputs.moonAltDeg)), 0, 1);

  const clearFcst = clamp(
    1.0 - (0.7 * inputs.cloudLowConsensus + 0.3 * inputs.cloudTotalConsensus),
    0,
    1,
  );

  let clear: number;
  let confidence: number;
  if (inputs.satelliteClearNow === null) {
    clear = clearFcst;
    confidence = (1.0 - inputs.cloudModelSpread) * 0.85;
  } else {
    clear = 0.5 * clearFcst + 0.5 * inputs.satelliteClearNow;
    confidence = (1.0 - inputs.cloudModelSpread) * 1.0;
  }

  const score = inputs.ovalVisibleProb * darkness * moonFactor * clear;

  let verdict: Verdict;
  if (score >= 0.3) {
    verdict = "Likely";
  } else if (score >= 0.1) {
    verdict = "Possible";
  } else {
    verdict = "Unlikely";
  }

  const factors: Record<Limiter, number> = {
    Daylight: darkness,
    Oval: inputs.ovalVisibleProb,
    CloudObserved: inputs.satelliteClearNow ?? 1.0,
    CloudForecast: clearFcst,
    Moon: moonFactor,
  };

  const best = Math.min(...Object.values(factors));
  const dominantLimiter =
    LIMITER_ORDER.find((k) => factors[k] === best) ?? "Oval";

  return { score, verdict, confidence, dominantLimiter };
}

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180;
}

File: crates/helio-core/src/golook.rs
//! §7.1 — "Go Look" score (scalar inputs ONLY — §4.1 invariant)

#[derive(Debug, Clone, Copy)]
pub struct GoLookInputs {
    pub oval_visible_prob: f64,
    pub sun_alt_deg: f64,
    pub moon_alt_deg: f64,
    pub moon_illum_frac: f64,
    pub cloud_total_consensus: f64,
    pub cloud_low_consensus: f64,
    pub cloud_model_spread: f64,
    pub satellite_clear_now: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Verdict {
    Likely,
    Possible,
    Unlikely,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Limiter {
    Daylight,
    Oval,
    CloudObserved,
    CloudForecast,
    Moon,
}

#[derive(Debug, Clone, Copy)]
pub struct GoLookScore {
    pub score: f64,
    pub verdict: Verdict,
    pub confidence: f64,
    pub dominant_limiter: Limiter,
}

/// Darkness factor for aurora visibility.
///
/// PINNED: clamp((−6 − sun_alt_deg) / 12, 0, 1)
/// Vectors: contracts/fixtures/vectors/golook.json, function "darkness_factor" (1e-9).
pub fn darkness_factor(sun_alt_deg: f64) -> f64 {
    let factor = (-6.0 - sun_alt_deg) / 12.0;
    factor.clamp(0.0, 1.0)
}

/// The on-device "go look" score. v1.0 heuristic.
///
/// PINNED semantics — see contracts/wasm-api/helio_core_api.rs.
/// Vectors: contracts/fixtures/vectors/golook.json, function "go_look" (1e-9 rel on score).
pub fn go_look(inputs: &GoLookInputs) -> GoLookScore {
    let darkness = darkness_factor(inputs.sun_alt_deg);
    let moon_factor = 1.0
        - 0.6
            * inputs.moon_illum_frac
            * inputs.moon_alt_deg.to_radians().sin().clamp(0.0, 1.0);
    let clear_fcst = (1.0
        - (0.7 * inputs.cloud_low_consensus + 0.3 * inputs.cloud_total_consensus))
        .clamp(0.0, 1.0);

    let (clear, confidence, cloud_observed_limiter) = match inputs.satellite_clear_now {
        None => (clear_fcst, (1.0 - inputs.cloud_model_spread) * 0.85, 1.0),
        Some(sat) => (
            0.5 * clear_fcst + 0.5 * sat,
            1.0 - inputs.cloud_model_spread,
            sat,
        ),
    };

    let score = inputs.oval_visible_prob * darkness * moon_factor * clear;
    let verdict = if score >= 0.30 {
        Verdict::Likely
    } else if score >= 0.10 {
        Verdict::Possible
    } else {
        Verdict::Unlikely
    };

    // Argmin with contract tie order: Daylight, Oval, CloudObserved,
    // CloudForecast, Moon. Strict comparisons preserve earlier ties.
    let mut best_value = darkness;
    let mut dominant_limiter = Limiter::Daylight;
    for (limiter, value) in [
        (Limiter::Oval, inputs.oval_visible_prob),
        (Limiter::CloudObserved, cloud_observed_limiter),
        (Limiter::CloudForecast, clear_fcst),
        (Limiter::Moon, moon_factor),
    ] {
        if value < best_value {
            best_value = value;
            dominant_limiter = limiter;
        }
    }

    GoLookScore {
        score,
        verdict,
        confidence,
        dominant_limiter,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_darkness_factor_day() {
        // Sun at +10° => full daylight => factor 0
        assert!((darkness_factor(10.0) - 0.0).abs() < 1e-9);
    }

    #[test]
    fn test_darkness_factor_night() {
        // Sun at -20° => full darkness => factor 1
        assert!((darkness_factor(-20.0) - 1.0).abs() < 1e-9);
    }

    #[test]
    fn test_darkness_factor_twilight() {
        // Sun at -12° => midpoint of ramp => factor 0.5
        assert!((darkness_factor(-12.0) - 0.5).abs() < 1e-9);
    }
}
